### PR TITLE
Use messagingReadyCallback in openSession/changeUser

### PR DESCRIPTION
**Pull request by ghost**
This should ensure that Braze triggers are synced before signaling that we're ready for use.

cc: @froodian @bucimis

### CODE REVIEW

**Review by Bucimis — APPROVED** [review]
appears fine from a cursory standpoint

**Comment by ladanazita** [comment]
Moving this in here from email for code-review/testing. Hunter mentions his impression is the solution may miss some cases.
 

> During the Braze destination initialization with Segment's analytics.js, analytics.js will:
> - if the user is anonymous, fire the ready() callback after Braze messages have been downloaded
> - if the user is identified, fire the ready() callback after Braze messages have been downloaded for the named user
>  
> This may miss some cases.  For example, if the identify call (and thus changeUser()) happens after the first ready() time has passed.

File: lib/index.js [hunk 1]
```diff
@@ -146,16 +147,16 @@ Appboy.prototype.initializeV2 = function(customEndpoint) {
...
-  
-  if (options.automaticallyDisplayMessages) window.appboy.display.automaticallyShowNewInAppMessages();
-  if (userId) window.appboy.changeUser(userId);
-  
-  window.appboy.openSession();
 
-  this.load('v2', this.ready);
+  if (options.automaticallyDisplayMessages) window.appboy.display.automaticallyShowNewInAppMessages();
+  if (userId) {
+    window.appboy.changeUser(userId, window.appboy.openSession(self.load('v2', self.ready)));
```
Comment (ucarion) [inline lib/index.js]: I think there's a minor oversight here.

`changeUser`'s second argument should be a function, no? Because the code as written, we'd be passing whatever `openSession` returns. I think we want to pass a function which in turn calls `openSession`.

A similar note for what we're passing to `openSession` -- we're passing `self.load(...)`, which isn't the right type.

In other words, I *think* we want to do:

```js
window.appboy.changeUser(userId, function () {
  window.appboy.openSession(function () {
    self.load('v2', self.ready);
  });
});
```

With the code as written, we'd evaluate `self.load(...)` as the *first* thing to happen, rather than the *last* thing.

File: lib/index.js [hunk 2]
```diff
@@ -146,16 +147,16 @@ Appboy.prototype.initializeV2 = function(customEndpoint) {
...
+  } else {
+    window.appboy.openSession(self.load('v2', self.ready));
```
Comment (ucarion) [inline lib/index.js]: See note above -- this code passes `self.load(...)`'s return value to `openSession`. That seems like a mistake?

**Review by ucarion — CHANGES_REQUESTED** [review]
Thanks so much for taking the time to open this PR! I've got two broad comments:

1. This code updates `initializeV2`, but do we not want to apply the same change to `initializeV1`?

2. I think there are some callback-related errors in this code. Let me know what you think.

:)

File: lib/index.js (hunk 1 above)
Comment (ghost) [reply lib/index.js]: Yep -- thanks a bunch for catching this.

**Comment by ghost** [comment]
re: v1, unfortunately, that version of the Appboy Web SDK doesn't support the `messagingReadyCallback`. In general, I think we should encourage people to use v2 as v1 is a pretty outdated version of the Web SDK.

**Comment by Bucimis** [comment]
@ladanazita gotcha, just to clarify, are you referring to the existing race condition where sometime after init, users call `identify()`/`changeUser()` and try to trigger messaging before `messagingReady` is called for the new user? Any other potential edges to be aware of?

Agreed that there are probably a few other cases where utilizing `messagingReady` would be useful, but not sure there's an easy way to work those in with existing Segment callbacks. We were mainly hoping to address the IBM-specific `page()` race condition and make things incrementally better.

**Comment by ladanazita** [comment]
@Bucimis Yup, this is my understanding of one of the edge-cases. After the call we had today, it sounds like there isn't a comprehensive list of edge-cases we can test against. IBM provided the one I posted above, but still mentioned potentially having others to consider.

They are testing some cases now and providing flows to verify this solution against. It does sound like they will need to consider the one mentioned above.

**Comment by hmedney** [comment]
I created a local build by combining the `analytics.js` produced for our test write key and the contents of the PR.  What I'm seeing is at this point in the lifecycle, inside `Appboy.prototype.initializeV2`, `window.appboy` is basically a placeholder with stubbed-out functions.

The call to `changeUser()` in the PR is actually calling this stub:
```js
(function anonymous() {
   return function changeUser(){appboyQueue.push(arguments); return true}
})
```
...which is created [here](https://github.com/segment-integrations/analytics.js-integration-appboy/blob/master/lib/index.js#L71).  Of course, the stub doesn't have any notion of the messaging ready callback.

From what I can gather, Segment stubs-out the Braze integration and its functions, enabling the browser to call them early.  Each invocation gets queued while the real js (Braze SDK) is downloaded.  Once downloaded, I'm guessing the queued invocations get played back.

So...maybe the change needs to happen after the real library is loaded but before the queued functions get played back?

**Comment by ghost** [comment]
`appboyQueue` is part of our [recommended integration snippet](https://github.com/appboy/appboy-web-sdk), and as you noted, it’s meant for playing back the interface calls after the Braze SDK gets downloaded. 

If the Braze SDK wasn’t loaded at this point, the method call gets queued, along with its arguments, and then replayed once Braze is finished loading, so I think we should be fine there (i.e. we'll queue up changeUser, along with the callback, and once the Braze SDK is loaded, replay that call).

Note that the network call to fetch data is separate from loading the Braze SDK, and only happens after the Braze SDK is loaded (i.e. when we invoke `changeUser` or `openSession`).

Looking at other parts of the integration, it does seem like there's a few places where we call `changeUser` followed by `logCustomEvent` -- there's room for improvement there to introduce logic that when we're on v2 (i.e. we have access to the `messagingReadyCallback`), we can pass that `logCustomEvent` as a callback instead, so that it only gets logged after we've synced triggers (if we're actually changing users). 

I think that coupled with the changes to the v2 initialization code should resolve IBM's issue, and I'm happy to drop that in this PR. 

We could theoretically also provide a `messagingReadyCallback` parameter to the `identify` interface so that it gets passed into `changeUser`. This would help alleviate issues from integrations calling identify and then immediately after logging a custom event that's meant to trigger an IAM based on newly synced triggers. Not sure if that's of immediate urgency though (as far as I understand, IBM's main issue was with `page`).

**Comment by hmedney** [comment]
Hi @viparthasarathy, thanks for the response.  My concern is even though the calls to the `changeUser()` stub will get replayed, it won't affect the initial Segment `ready()` callback - Segment's `ready()` will have fired already by the time `changeUser(user, messagingReadyCallback)` is replayed to the SDK, and thus we don't have a suitable callback to know when messages are downloaded.

One suggestion was to enhance the Braze SDK with new top-level function `messagingReady(callback)` that behaves like `messagingReadyCallback` callback on session init, but be callable any time (similar to Segment's analytics.ready()).
 
For example,
 ```js
appboy.messagingReady(function() {
    appboy.logCustomEvent("Some event we want to fire when we know in-app messages have been downloaded")
})
```
 
If `messagingReady()` behaved as follows:
- call the callback immediately if messages don't need to be downloaded
- call the callback immediately if messages have been downloaded
- if messages are being downloaded, let's say for a new session, then call the callback when messages have been downloaded
 
Then the appboy integration with Segment can simply wrap each call to appboy.logCustomEvent() in `appboy.messagingReady()`.  This should queue events until messages are available and fire events immediately if messages are already available.  It would also gracefully handle multiple Segment identifies in a single page.  This should cover most if not all the cases.

**Comment by ghost** [comment]
Hey @hmedney, thanks for the response.

From my understanding, the changes to the v2 initialization code here should prevent Segment from being ready _until_ the `changeUser`call gets replayed as we've now passed `self.load('v2', self.ready)` into the `messagingReadyCallback`. I might be missing something here though, and it would be good to have someone from Segment's side chime in. 

I'm not sure I see a strong use case for a new `messagingReady` function as opposed to getting this integration to just better utilize the `messagingReadyCallback` for `openSession` / `changeUser`, but it'd be good to have concrete examples of where you think the latter wouldn't really do what you need it to do.

**Comment by hmedney** [comment]
Hi @viparthasarathy, what seems to be happening is the callback isn't being called while queuing calls, e.g.
screenshot from 2018-07-25 13-03-44

The 2nd and 3rd `debugger` statements are never reached since the first callback doesn't fire, and thus `self.load('v2', self.ready);` is never reached (this line loads the actual SDK code I believe by injecting a script tag)

...:chicken: and :egg: problem

**Comment by ucarion** [comment]
Hey! So indeed, there's a bug in this PR, and my initial suggestion was indeed a bit wrong. :)

`this.load` is the method that will insert a Appboy's script tag onto the page. Until `this.load` is called, calling methods on `window.appboy` will be queued indefinitely. So we just to change the order we do things.

Right now, we do things like this:

```js
    window.appboy.changeUser(userId, function() {
      window.appboy.openSession(function() {
        self.load('v2', self.ready);
      });
    });
```

That won't work -- until `self.load` happens, the `changeUser` callback has no hope of ever being called. Instead, we should do this:

```js
self.load('v2', function() {
  window.appboy.changeUser(userId, function() {
    window.appboy.openSession(function() {
      self.ready();
    });
  });
});
```

**Comment by hmedney** [comment]
@ucarion that looks promising - will give it a try!

**Comment by hmedney** [comment]
One item I wanted to highlight on the revised solution (wrapping `page` and `track` calls behind `changeUser(user, callback)`) is I've been seeing an issue in the current integration code:

```js
Appboy.prototype.page = function(page) {
  var settings = this.options;
  if (!settings.trackAllPages && !settings.trackNamedPages) return;
  if (settings.trackNamedPages && !page.name()) return;

  var userId = page.userId();
  var pageEvent = page.track(page.fullName());
  var eventName = pageEvent.event();
  var properties = page.properties();

  window.appboy.changeUser(userId);
  window.appboy.logCustomEvent(eventName, properties);
};
```

When Braze SDK logging is enabled, the following is reported on the console:
```
Appboy SDK Error: changeUser requires a non-empty userId.
```
...which suggest userId is null in the `changeUser()` call.  The user should be known as they've been identified from a previous `analytics.identify()` call.

**Comment by ladanazita** [comment]
@hmedney Caught another bug here. `page.userId()` on the client will never return a `userId`, regardless of the `identify` being called and setting the `userId` on the cache, because analytics.js does not set `userId` on the `page` facade.

We will need to update this to retrieve `userId` with `analytics.user().id()`

File: lib/index.js [hunk 3]
```diff
@@ -276,13 +300,25 @@ Appboy.prototype.page = function(page) {
   if (!settings.trackAllPages && !settings.trackNamedPages) return;
   if (settings.trackNamedPages && !page.name()) return;
 
-  var userId = page.userId();
+  var userId = this.analytics.user().id();
```
Comment (hmedney) [inline lib/index.js:311]: @ladanazita @ucarion @viparthasarathy do cases exist where `this.analytics.user()` is undefined?  I remember having to guard against that in some cases

File: lib/index.js [hunk 4]
```diff
@@ -258,8 +270,20 @@ Appboy.prototype.track = function(track) {
     delete properties[key];
   }, reserved);
 
-  window.appboy.changeUser(userId);
-  window.appboy.logCustomEvent(eventName, properties);
+  if (this.supportsMessagingReadyCallback) {
```
Comment (hmedney) [inline lib/index.js:281]: @ladanazita @ucarion @viparthasarathy should the same userId code from `prototype.page` go here as well?

File: lib/index.js [hunk 5]
```diff
@@ -146,16 +148,26 @@ Appboy.prototype.initializeV2 = function(customEndpoint) {
...
   window.appboy.initialize(options.apiKey, config);
-  
-  if (options.automaticallyDisplayMessages) window.appboy.display.automaticallyShowNewInAppMessages();
-  if (userId) window.appboy.changeUser(userId);
-  
-  window.appboy.openSession();
 
-  this.load('v2', this.ready);
+  if (options.automaticallyDisplayMessages) window.appboy.display.automaticallyShowNewInAppMessages();
+  if (userId) {
```
Comment (hmedney) [inline lib/index.js:164]: @ladanazita @ucarion @viparthasarathy try inverting the userId check and load:

```
  self.load('v2', function() {
    var userId = self.analytics.user().id();
    if (userId) {
      window.appboy.changeUser(userId, function() {
        window.appboy.openSession(function() {
          self.ready();
        });
      });
    } else {
      window.appboy.openSession(function() {
        self.ready();
      });
    }
  });
};
```
(I think this was the main reason the test earlier didn't work as expected)

**Comment by hmedney** [comment]
@viparthasarathy when `changeUser()` is called, and the userId hasn't changed, the `messagingReadyCallback` is called immediately.  It doesn't seem to factor-in any in-flight requests.

Here's the relevant minified portion of the Braze SDK:
```js
b.I.prototype.Se = function(a, c, d) {
  var e = this.H.U();
  if (e !== a) {  // comparing current user id to user id in changeUser()
    this.Y.Ti();
    this.H.lj(a);
    if (null != e) {
      for (e = 0; e < c.length; e++) c[e].clearData(!0);
      this.A.Ua(b.h.B.D.re);
      this.A.Ua(b.h.B.D.oc);
    }
    this.A.Ua(b.h.B.D.oe);
    c = this.A.Aa(b.h.B.D.qd);
    null != c && 0 < c.length && this.jc().Xc(c[0], c[1], c[2], c[3]);
    this.Xd(null, d);
    f.b.info('Changed user to "' + a + '".');
  } else
    // user ids are equal;
    // call the messagingReadyCallback immediately if provided and is a function
    'function' === typeof d
      ? (f.b.info('Current user is already "' + a + '". Executing networkCallCompleteCallback.'), d()) // d is the messagingReadyCallback
      : f.b.info('Current user is already "' + a + '". Doing nothing.');
};
```
(added comments to explain what seems to be happening based on tracing the code in the debugger)

**Comment by hmedney** [comment]
Based on the above, wrapping the Segment `track` and `page` calls behind `changeUser(userId, messagingReadyCallback)` doesn't get us there since `messagingReadyCallback` is called immediately if `userId` is the same as the current Braze user id, e.g. recently set by `analytics.identify(userId)`.

On the plus side, if `userId` and `self.load()` are inverted as suggested in the review comment, then the the desired flow will work *if* the identify call is made early enough to take effect before the SDK finishes downloading.

## Flows
Below are 2 flows to illustrate a working and non-working scenario.  Each row is in format:
- [timestamp] [caller] function called / what happened

### Identify call is made early: :heavy_check_mark:

1. 00:00 [page] `analytics.load()`
1. 00:00 [page] `analytics.identify("IBMId-123")`
1. 00:00 [page] `analytics.ready(segmentReadyCallback)`
1. 00:00 [analytics] `this.load("v2", integrationLoaded)`
1. (analytics finishes loading Braze/AppBoy SDK)
1. 00:01 [integrationLoaded] `appboy.changeUser("IBMid-123", messagingReadyCallback1)`
1. 00:02 [appboy] Begin downloading messages for "IBMId-123"
1. 00:02 [appboy] Finished downloading messages for "IBMId-123"
1. 00:02 [messagingReadyCallback1] notify `analytics` SDK is ready
1. 00:02 [segmentReadyCallback] `analytics.page()`
1. 00:02 [analytics] `appboy.changeUser("IBMid-123", messagingReadyCallback2)` (`messagingReadyCallback2` resolves immediately)
1. 00:02 [messagingReadyCallback2] `appboy.logCustomEvent("Loaded a Page")`
1. 00:02 [appboy] In-app message for IBMId-123 is displayed :heavy_check_mark:

### Identify call is made late: :x:
1. 00:00 [page] `analytics.load()`
1. 00:00 [page] `analytics.ready(segmentReadyCallback)`
1. 00:00 [analytics] `this.load("v2", integrationLoaded)`
1. (analytics finishes loading Braze/AppBoy SDK)
1. 00:01 [analytics] `appboy.openSession(messagingReadyCallback1)`
1. 00:02 [appboy] Begin downloading messages for anonymous user
1. 00:02 [appboy] Finished downloading messages for anonymous user
1. 00:02 [messagingReadyCallback1] notify `analytics` Braze SDK is ready
1. 00:02 [segmentReadyCallback] `analytics.identify("IBMId-123")`
1. 00:02 [analytics] `appboy.changeUser("IBMid-123")` (no callback)
1. 00:02 [appboy] Begin downloading messages for "IBMId-123"
1. 00:02 [segmentReadyCallback] `analytics.page()`
1. 00:02 [analytics] `appboy.changeUser("IBMid-123", messagingReadyCallback2)` (`messagingReadyCallback2` resolves immediately since the userId is unchanged)
1. 00:02 [messagingReadyCallback2] `appboy.logCustomEvent("Loaded a Page")`
1. 00:02 [appboy] In-app message for anonymous user is displayed :x: 
1. 00:03 [appboy] Finished downloading messages for "IBMId-123"

## Handling late identify and other edge cases
To support the late identify case, we need `changeUser(userId, messagingReadyCallback)` to call `messagingReadyCallback` when the following are met:
- call immediately if messages don't need to be downloaded
- call immediately if messages have been downloaded
- if messages are being downloaded, call when messages have been downloaded

To maintain compatibility, perhaps this could be a new option passed-in to `changeUser()`

**Comment by ghost** [comment]
Hey @hmedney, thanks for that thorough breakdown. We were discussing this internally and we came to similar conclusions regarding the flow you outlined. I think the optimal solution here would be to reconsider our usage of `changeUser` throughout this integration -- IMO, there should only be one interface for calling `appboy.changeUser` (and it seems like `analytics.identify` is the most reasonable candidate), and that interface should have a way of passing a `messagingReadyCallback` so that we can do things after we’ve synced triggers for that user. I’d imagine the flow for that would look something like:

```
analytics.load();
analytics.ready(function() {
  analytics.identify(brazeUserId, function() {
     analytics.page();
  });
});
```

Alternatively, as you noted, we could try to find some way to ensure that the userId value is populated before the Braze SDK gets downloaded. Right now, it seems like one potential way is what you suggested, though at that point, we’re reliant on a race condition panning out the way we want it to, which isn’t ideal. It’d be nicer if there was some way to pass in the userId to the initialization of Braze. I’d imagine that’d just look something like this:

```
analytics.load();
analytics.ready({brazeUserId: 'userId'}, function() {
  analytics.identify(brazeUserId); // triggers already pulled on braze side.
  analytics.page();
});
```

I’m in favor of the former approach as I think it’ll be much more flexible to other use cases. With the latter, I can see issues if you try to identify as a different user after loading and try to immediately trigger an IAM specific to them.

Curious as to Segment's thoughts here @ucarion.

File: lib/index.js (hunk 3 above)
Comment (ucarion) [reply lib/index.js:311]: There should be no need to guard against `user()` being undefined in this case. In `initialize`, a guard would be required, however.

File: lib/index.js (hunk 4 above)
Comment (ucarion) [reply lib/index.js:281]: Yes, it should.

**Comment by ucarion** [comment]
I realize we're discussing an alternative solution in email, but here's a contingency plan we could deliver:

In `identify()`, we could alter this integration to pass a global variable as a callback to `changeUser`, like so:

```js
window.appboy.changeUser(userId, function () {
  var callback = window.__hack_segment_appboy_change_user_cb;
  if (callback) {
    callback();
  }
});
```

Then, IBM could hook into this callback to ensure that `page()` calls don't happen until IAM are ready.

```js
// `analytics.ready` is still required, because the `changeUser` call might fire at initialization-time.
// `analytics.ready` hooks into that first `changeUser` call's callback.
analytics.ready(function () {
  // Prepare a callback for the second `changeUser` call.
  // So long as Braze is enabled on the page, this callback is guaranteed to be called.
  // But if Braze is disabled, then IBM will need to call `analytics.page` differently.
  window.__hack_segment_appboy_change_user_sb = function () {
    analytics.page(...);  
  });

  analytics.identify(...);
});
```

cc @hmedney @viparthasarathy

**Comment by hmedney** [comment]
@ucarion interesting idea.  Does a corresponding place exist for calling `changeUser()` during rehydration?

**Comment by ghost** [comment]
That sounds like a viable solution to me that should cover all the bases here, as long as the flow is always as above, i.e. hook into ready, set up the callback, then identify. We should always have the appropriate synced data (either ready covered it or identify covers it).

File: lib/index.js (hunk 4 above)
Comment (ghost) [reply lib/index.js:281]: Gotcha, will update. I'm questioning if changeUser is even really needed here (pretty sure it's always a no-op because either it was cached and we've called changeUser at initialization time or we've called identify and called changeUser in there), but don't want to get sidetracked.

File: lib/index.js (hunk 5 above)
Comment (ghost) [reply lib/index.js:164]: Going to assume this isn't necessary anymore as we'd want to keep the `.identify` call inside of the `.ready` callback anyways with the suggested global callback solution.

File: lib/index.js [hunk 6]
```diff
@@ -193,7 +213,7 @@ Appboy.prototype.identify = function(identify) {
   var phone = identify.phone();
   var traits = clone(identify.traits());
 
-  window.appboy.changeUser(userId);
+  window.appboy.changeUser(userId, window.__hack_segment_appboy_change_user_sb);
```
Comment (ghost) [inline lib/index.js:216]: `changeUser` only calls `messagingReadyCallback` if it's a function, so we can just pass this variable in.

**Comment by hmedney** [comment]
@viparthasarathy @ucarion please take a look at an implementation of the solution [here](https://github.com/hmedney/analytics.js-integration-appboy/blob/global-messaging-ready-callback/lib/index.js)

Basically it keeps track of the callback in a `window.appboy` var and provides a new function called `_messagingReady(callback)`.  Calls to `logCustomEvent()` can be wrapped as follows:

```js
this._messagingReady(function() {
  window.appboy.logCustomEvent(eventName, properties);
})
```
When the messages have been downloaded, the callback is invoked.  In testing simple pages, it seems to solve the race condition and make the integration more predictable.

For example, if a page instruments as follows:
```js
analytics.load("{writeKey}");
analytics.identify("{userId}")
analytics.page();
```
The above change will ensure all the Braze in-app messages have been downloaded before the page custom event is called on the SDK.
